trio/_core/_asyncgens.py

```diff
@@ -61,7 +70,7 @@ def finalize_in_trio_context(agen, agen_name):
                 # have hit it.
                 self.trailing_needs_finalize.add(agen)
 
-        def finalizer(agen):
+        def finalizer(agen: AsyncGenerator[Any, Any]) -> None:
```

Comment:
To fix the `closer.send(None)` error, change this to `types.AsyncGeneratorType` (and probably revert the same change in `_util.name_asyncgen`. The subtle difference between them is that [`collections.abc.AsyncGenerator`](https://github.com/python/typeshed/blob/75c71f324ec564f2c6eaeee81f6b9618db5fb98c/stdlib/typing.pyi#L477) defines the methods as just returning an `Awaitable`, while [`types.AsyncGeneratorType`](https://github.com/python/typeshed/blob/75c71f324ec564f2c6eaeee81f6b9618db5fb98c/stdlib/types.pyi#L379C17-L379C17) corresponds specifically to the type produced by `async def` functions and defines the methods as returning coroutines. 

This finalizer function is called by the interpreter, we should be safe in assuming that they're Python-defined generators and not some other class. Though, the stubs for `set_asyncgen_hooks()` use the regular ABC type...

Reply:
Changed in 0205c8f